src/human_encoding/mod.rs

```diff
@@ -0,0 +1,55 @@
+// Rust Simplicity Library
+// Written in 2023 by
+//   Andrew Poelstra <rust-simplicity@wpsoftware.net>
+//   Christian Lewe <clewe@blockstream.com>
+//   Sanket Kanjalkar <sanket1729@gmail.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! The Simplicity Human-Readable Encoding
+//!
+//! This module provides the ability to decode and encode [`crate::NamedCommitNode`]s
+//! in a human-readable format.
+//!
+
+mod named_node;
+
+use crate::jet::Jet;
+use crate::node::CommitNode;
+
+use std::collections::HashMap;
+use std::str;
+use std::sync::Arc;
+
+pub use self::named_node::NamedCommitNode;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Forest<J: Jet> {
+    programs: HashMap<Arc<str>, Program<J>>,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Program<J: Jet> {
+    root: Arc<NamedCommitNode<J>>,
+}
```

Comment:
cf2ea20bbbf0297347b132d7c1b382e88f8997e4: We should think about how we handle programs of other node types. So far, we use `&Node` or `Arc<Node>` in methods like `BitMachine::for_program()`.

Reply:
Well, right now we take a `RedeemNode` in the bit machine because that's the only node type that is guaranteed to have witness values, complete types, and disconnected expressions. If we add another such node type we'll need to add a trait describing this.

Reply:
Nothing in my PR stack involves a `NamedRedeemNode` and I haven't found a use for one ... I think, if you want to execute a program, you should just convert it to `RedeemNode` beforehand.